### PR TITLE
Rdcc 1640

### PR DESCRIPTION
**Pull request by KOTLAPRA23**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/RDCC-1640

### Change description ###

feat:audit functionality implemented for profile sync service

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ X] No
```

### CODE REVIEW

File: src/test/java/uk/gov/hmcts/reform/profilesync/schedular/UserProfileSyncJobSchedulerTest.java [hunk 1]
```diff
@@ -2,85 +2,115 @@
...
-import uk.gov.hmcts.reform.profilesync.domain.SyncJobAudit;
+import uk.gov.hmcts.reform.profilesync.domain.ProfileSyncAudit;
 import uk.gov.hmcts.reform.profilesync.domain.SyncJobConfig;
+import uk.gov.hmcts.reform.profilesync.repository.ProfileSyncAuditRepository;
 import uk.gov.hmcts.reform.profilesync.repository.SyncConfigRepository;
-import uk.gov.hmcts.reform.profilesync.repository.SyncJobRepository;
 import uk.gov.hmcts.reform.profilesync.service.ProfileSyncService;
 
 public class UserProfileSyncJobSchedulerTest {
+    //mocked as its an interface
```
Comment (AdilOoze) [inline src/test/java/uk/gov/hmcts/reform/profilesync/schedular/UserProfileSyncJobSchedulerTest.java:20]: I don't think this comment is required anymore if you removed `syncJobRepository`

Comment (KOTLAPRA23) [reply src/test/java/uk/gov/hmcts/reform/profilesync/schedular/UserProfileSyncJobSchedulerTest.java:20]: ProfileSyncAuditRepository also interface so comment is valid one only

**Comment by shashank-rastogi** [comment]
Please ensure mutation test coverage is above threshold

**Comment by KOTLAPRA23** [comment]
> Please ensure mutation test coverage is above threshold

yes .it is 93 % and above the threshold

**Comment by shashank-rastogi** [comment]
looks like class UserProfileResponse is not used anywhere.. if its not used then probably u can remove it

File: src/main/java/uk/gov/hmcts/reform/profilesync/schedular/UserProfileSyncJobScheduler.java [hunk 2]
```diff
@@ -41,55 +41,59 @@
...
+            log.error("{}::Sync Batch Job Failed::{}", loggingComponentName, e.getErrorMessage());
+            syncAudit.setSchedulerStatus("fail");
+            syncAudit.setSchedulerStartTime(startTime);
+            profileSyncAuditRepository.save(syncAudit);
 
         }
     }
 
-    private String getLastBatchFailureTimeInHours(LocalDateTime lastSuccessBatch) {
+    public String getLastBatchFailureTimeInHours(LocalDateTime lastSuccessBatch) {
```
Comment (shashank-rastogi) [inline src/main/java/uk/gov/hmcts/reform/profilesync/schedular/UserProfileSyncJobScheduler.java:96]: the method name is wrong. this method is to calculate the time difference from last successful run and not for batch failure

File: src/main/java/uk/gov/hmcts/reform/profilesync/service/impl/ProfileSyncServiceImpl.java [hunk 3]
```diff
@@ -95,49 +95,47 @@ public String getS2sToken() {
     }
 
 
-    public List<IdamClient.User> getSyncFeed(String bearerToken, String searchQuery) throws UserProfileSyncException {
+    public Set<IdamClient.User> getSyncFeed(String bearerToken, String searchQuery)throws UserProfileSyncException {
         Map<String, String> formParams = new HashMap<>();
         formParams.put("query", searchQuery);
 
-        List<IdamClient.User> updatedUserList = new ArrayList<>();
+        Set<IdamClient.User> updatedUserList = new HashSet<>();
```
Comment (shashank-rastogi) [inline src/main/java/uk/gov/hmcts/reform/profilesync/service/impl/ProfileSyncServiceImpl.java:102]: updatedUserList is a set and not a list

File: src/main/java/uk/gov/hmcts/reform/profilesync/service/impl/ProfileUpdateServiceImpl.java [hunk 4]
```diff
@@ -38,81 +41,76 @@
...
                         .firstName(user.getForename())
                         .lastName(user.getSurname())
-                        .idamStatus(idamStatusResolver().get(status) != null ? idamStatusResolver().get(status).name() : IdamStatus.SUSPENDED.name())
+                        .idamStatus(resolveIdamStatus(sb))
                         .build();
 
                 try {
-
-                    syncUser(bearerToken, s2sToken, user.getId(), updatedUserProfile);
+                    profileSyncAuditDetails.add(syncUser(bearerToken, s2sToken, user.getId(), updatedUserProfile,syncAudit));
```
Comment (shashank-rastogi) [inline src/main/java/uk/gov/hmcts/reform/profilesync/service/impl/ProfileUpdateServiceImpl.java:65]: this is not the best practice. SyncUser is the main method which syncs the user status with User profile. However it is called an argument to the audit method which makes it hard to understand the input being passed and also kind of give impression that profileSyncAuditDetails.add() is doing main business logic
It makes it bit more difficult to understand and maintain

File: src/main/java/uk/gov/hmcts/reform/profilesync/domain/ProfileSyncAudit.java [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    @Column(name = "scheduler_start_time")
+    private LocalDateTime schedulerStartTime;
+
+    @CreationTimestamp
+    @Column(name = " scheduler_end_time")
+    private LocalDateTime schedulerEndTime;
+
+    @OneToMany(targetEntity = ProfileSyncAuditDetails.class, cascade = CascadeType.ALL)
+    @JoinColumn(name = "id", referencedColumnName = "scheduler_id")
+    private List<ProfileSyncAuditDetails> profileSyncAuditDetails = new ArrayList<>();
```
Comment (shashank-rastogi) [inline src/main/java/uk/gov/hmcts/reform/profilesync/domain/ProfileSyncAudit.java:47]: why is this required here? you have already linked it by using Many to one relationship in ProfileSyncAuditDetailsId

File: src/integrationTest/java/uk/gov/hmcts/reform/profilesync/RunProfileSyncAuditJobIntTest.java [hunk 6]
```diff
@@ -10,24 +11,25 @@
...
+import uk.gov.hmcts.reform.profilesync.repository.ProfileSyncAuditDetailsRepository;
+import uk.gov.hmcts.reform.profilesync.repository.ProfileSyncAuditRepository;
 import uk.gov.hmcts.reform.profilesync.repository.SyncConfigRepository;
-import uk.gov.hmcts.reform.profilesync.repository.SyncJobRepository;
 import uk.gov.hmcts.reform.profilesync.schedular.UserProfileSyncJobScheduler;
 
 @Slf4j
 @RunWith(SpringIntegrationSerenityRunner.class)
-public class RunProfileSyncJobTest extends AuthorizationEnabledIntegrationTest {
+public class RunProfileSyncAuditJobIntTest extends AuthorizationEnabledIntTest {
```
[the remark this answers is not in the document]
Comment (KOTLAPRA23) [reply src/integrationTest/java/uk/gov/hmcts/reform/profilesync/RunProfileSyncAuditJobIntTest.java:23]: to append Int to the test class renamed it

File: src/main/resources/db/migration/V1_4__add_sync_audit_new_tables copy.sql [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+
+create table profile_sync_audit_details(
+id bigint,
+user_identifier varchar(50),
+status_code bigint,
+error_description varchar(1024),
+created_timestamp timestamp,
+constraint sync_audit_mapping_pk primary key (id,user_identifier)
+);
+
```
[the remark this answers is not in the document]
Comment (KOTLAPRA23) [reply src/main/resources/db/migration/V1_4__add_sync_audit_new_tables copy.sql:26]: done

File: src/main/java/uk/gov/hmcts/reform/profilesync/schedular/UserProfileSyncJobScheduler.java (hunk 2 above)
Comment (KOTLAPRA23) [reply src/main/java/uk/gov/hmcts/reform/profilesync/schedular/UserProfileSyncJobScheduler.java:96]: done

File: src/main/java/uk/gov/hmcts/reform/profilesync/service/impl/ProfileSyncServiceImpl.java (hunk 3 above)
Comment (KOTLAPRA23) [reply src/main/java/uk/gov/hmcts/reform/profilesync/service/impl/ProfileSyncServiceImpl.java:102]: yes modified from list to set

File: src/main/resources/db/migration/V1_4__add_sync_audit_new_tables copy.sql [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+create table profile_sync_audit(
+scheduler_id bigint,
+scheduler_start_time timestamp,
```
[the remark this answers is not in the document]
Comment (KOTLAPRA23) [reply src/main/resources/db/migration/V1_4__add_sync_audit_new_tables copy.sql:4]: in the sequence creation already mentioned  and it is taking serial id only.

File: src/main/java/uk/gov/hmcts/reform/profilesync/service/impl/ProfileUpdateServiceImpl.java (hunk 4 above)
Comment (KOTLAPRA23) [reply src/main/java/uk/gov/hmcts/reform/profilesync/service/impl/ProfileUpdateServiceImpl.java:65]: i have added the comment like what theline is doing to avode confusion. I have tried but no other way

File: src/integrationTest/java/uk/gov/hmcts/reform/profilesync/RunProfileSyncAuditJobIntTest.java [hunk 9]
```diff
@@ -39,14 +41,16 @@
...
-        SyncJobAudit syncJobAudit = syncJobRepository.findFirstByStatusOrderByAuditTsDesc("success");
-        assertThat(syncJobRepository.findAll()).isNotEmpty();
-        assertThat(syncJobAudit).isNotNull();
-        assertThat(syncJobAudit.getStatus()).isEqualTo("success");
+        ProfileSyncAudit profileSyncAudit = profileSyncAuditRepository.findFirstBySchedulerStatusOrderBySchedulerEndTimeDesc("success");
+        assertThat(profileSyncAuditRepository.findAll()).isNotEmpty();
+        assertThat(profileSyncAuditDetailsRepository.findAll()).isNotEmpty();
+        assertThat(profileSyncAudit).isNotNull();
+        assertThat(profileSyncAudit.getSchedulerStatus()).isEqualTo("success");
     }
```
[the remark this answers is not in the document]
Comment (KOTLAPRA23) [reply src/integrationTest/java/uk/gov/hmcts/reform/profilesync/RunProfileSyncAuditJobIntTest.java:54]: updating and adding new test.

File: src/main/java/uk/gov/hmcts/reform/profilesync/domain/ProfileSyncAudit.java (hunk 5 above)
Comment (KOTLAPRA23) [reply src/main/java/uk/gov/hmcts/reform/profilesync/domain/ProfileSyncAudit.java:47]: for composite key mapping  we need this and used  cascade  so that when main entity saves it automatically saves the sub entities associated . No need to add multiple save operation for each entity.

**Comment by KOTLAPRA23** [comment]
closing the PR since modifying the existing postgres script.

File: src/main/java/uk/gov/hmcts/reform/profilesync/service/impl/ProfileSyncServiceImpl.java (hunk 3 above)
Comment (shashank-rastogi) [reply src/main/java/uk/gov/hmcts/reform/profilesync/service/impl/ProfileSyncServiceImpl.java:102]: the name should be changed to updatedUserSet

File: src/main/resources/db/migration/V1_4__add_sync_audit_new_tables copy.sql (hunk 8 above)
[the remark this answers is not in the document]
Comment (shashank-rastogi) [reply src/main/resources/db/migration/V1_4__add_sync_audit_new_tables copy.sql:4]: the sequence does not use serial.. if u use "serial" you dont have to create a sequence in database
